### PR TITLE
feat(esl-event-listener): `ESLProxyUtils.decorate` target to create shared `debounced` | `throttled` subscriptions (POC)

### PR DESCRIPTION
**Pull request by ala-n**
Suggestion #1199 API implementation.

## The idea
`ESLEventUtils` will be extended with a new set of utilities to create an EventTarget-s with special behavior.

In bounds of the current PR, a new `ESLEventUtils.decorate` utility is created to create a cached EvenTarget with a decorated handler above the handler of the original EventTarget.

 `throttle` | `debounce` | `raf` or similar decoration functions can be used for handler decoration.
 Decoration happens per event, meaning different events will be processed with a separate decorated handler implementation.

## API:
1. Debounced resize with a default timeout 
```typescript
    // Instead @listen({event: 'resize', target: window})
    @listen({
      event: 'resize', 
      target: ESLEventUtils.decorate(window, debounce, 250)
    })
    onResize() { /* Debounced resize */ }
```
2. Throttled scroll wit a default timeout
```typescript
    @listen({
      event: 'scroll', 
      target: ESLEventUtils.decorate(window, throttle, 250)
    })
    onScroll() { /* Throttled scroll */ }
```

### CODE REVIEW

File: src/modules/esl-event-listener/core/targets/decorated.target.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  public override removeEventListener(event: any, callback: EventListener = event): void {
+    super.removeEventListener(event, callback);
+
+    if (!this.hasEventListener(event)) {
+      ESLEventListener.get(this, event).forEach((listener: ESLEventListener) => listener.unsubscribe());
+    }
+  }
+
+  /** @returns debounced handler */
+  protected get onResizeDebounced(): EventListener {
```
Comment (fshovchko) [inline src/modules/esl-event-listener/core/targets/decorated.target.ts:63]: Is method supposed to be named 'onResizeDebounced' ?

File: src/modules/esl-event-listener/test/targets/decorator.test.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    });
+  });
+
+  describe('ESLEventUtils.decorate attribute processing', () => {
+    const el = document.createElement('div');
+    const handler = jest.fn();
+    const dec = jest.fn(() => handler);
+    const decorated = ESLEventUtils.decorate(el, dec);
+
+    test('Creation does npt cause execution', () => expect(dec).not.toBeCalled());
```
Comment (dshovchko) [inline src/modules/esl-event-listener/test/targets/decorator.test.ts:33]:
```suggestion
    test('Creation does not cause execution', () => expect(dec).not.toBeCalled());
```

File: src/modules/esl-event-listener/test/targets/decorator.test.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        target: ESLEventUtils.decorate(window, debounce, DEFAULT_TIMEOUT)
+      }, fn);
+      jest.useFakeTimers();
+    });
+    afterAll(() => ESLEventUtils.unsubscribe(host));
+
+    beforeEach(() => fn.mockReset());
+    afterEach(() => jest.runAllTimers());
+
+    test('Proxy resize receive the resize event',  ()=> {
```
Comment (dshovchko) [inline src/modules/esl-event-listener/test/targets/decorator.test.ts:71]:
```suggestion
    test('Proxy resize receives the resize event',  ()=> {
```

File: src/modules/esl-event-listener/test/targets/decorator.test.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        target: ESLEventUtils.decorate(window, throttle, DEFAULT_TIMEOUT)
+      }, fn);
+      jest.useFakeTimers();
+    });
+    afterAll(() => ESLEventUtils.unsubscribe(host));
+
+    beforeEach(() => fn.mockReset());
+    afterEach(() => jest.runAllTimers());
+
+    test('Proxy scroll receive the scroll event',  ()=> {
```
Comment (dshovchko) [inline src/modules/esl-event-listener/test/targets/decorator.test.ts:120]:
```suggestion
    test('Proxy scroll receives the scroll event',  ()=> {
```

**Review by dshovchko — APPROVED** [review]

**Pull request merged**